lisa/tools/ethtool.py

```diff
@@ -0,0 +1,181 @@
+
+        self.device_name = device
+        self.enabled_features = []
+        for row in matched_features_info.group("value").splitlines():
+            feature_info = self.__feature_settings_pattern.match(row)
+            if not feature_info:
+                raise LisaException()
+            if "on" in feature_info.group("value"):
+                self.enabled_features.append(feature_info.group("name"))
+
+
+class Ethtool(Tool):
+    @property
+    def command(self) -> str:
+        return self._command
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        self._command = "ethtool"
+        self._device_list: List[str] = []
+        self._device_channel: List[DeviceChannel] = []
+        self._device_features: List[DeviceFeatures] = []
+
+    def _install(self) -> bool:
+        posix_os: Posix = cast(Posix, self.node.os)
+        posix_os.install_packages("ethtool")
+        return self._check_exists()
+
+    def _get_device_list(self) -> None:
+        if self._device_list:
+            return
+
+        netdirs = list(Path("/sys/devices/").rglob("*VMBUS*/**/net*"))
+        for netdir in netdirs:
+            devices = [e for e in netdir.iterdir() if e.is_dir()]
+            for device in devices:
+                if device.name in self._device_list:
+                    continue
+                self._device_list.append(device.name)
+
+    def get_device_channels_info(
+        self,
+        interface: str,
+    ) -> DeviceChannel:
+        for device in self._device_channel:
+            if device.device_name == interface:
+                return device
+
+        result = self.run(f"-l {interface}", shell=True)
+        if result.exit_code != 0:
+            raise LisaException(
+                f"ethtool -l {interface} command got non-zero"
+                f" exit code: {result.exit_code}"
+                f" error code: {result.stderr}"
+            )
+        device_channel_info = DeviceChannel(interface, result.stdout)
+        self._device_channel.append(device_channel_info)
+        return device_channel_info
+
+    def get_device_enabled_features(self, interface: str) -> DeviceFeatures:
+        for device in self._device_features:
+            if device.device_name == interface:
+                return device
+
+        result = self.run(f"-k {interface}", shell=True)
+        if result.exit_code != 0:
+            raise LisaException(
+                f"ethtool -k {interface} command got non-zero"
+                f" exit code: {result.exit_code}"
+                f" error code: {result.stderr}"
+            )
+        device_feature = DeviceFeatures(interface, result.stdout)
+        self._device_features.append(device_feature)
+
+        return device_feature
+
+    def get_all_device_channels_info(self) -> List[DeviceChannel]:
+        if self._device_channel:
```

Comment:
Based on current logic, if this method is called after one device is gotten, the list won't be completed.